crates/uv-git/src/git.rs

```diff
@@ -969,13 +973,28 @@ pub(crate) fn fetch(
 
     debug!("Performing a Git fetch for: {remote_url}");
     match strategy {
-        FetchStrategy::Cli => fetch_with_cli(repo, remote_url, &refspecs, tags),
+        FetchStrategy::Cli => {
+            fetch_with_cli(repo, remote_url, &refspecs, tags).or_else(|err| {
+                // If secondary refspecs are populated i.e. from [`GitReference::BranchOrTag`],
+                // then ignore the first error and try again; ideally we'd only retry if the
+                // error was due to a missing ref but this is already not the fast path
+                if secondary_refspecs.is_empty() {
+                    Err(err)
+                } else {
+                    fetch_with_cli(repo, remote_url, &secondary_refspecs, tags)
+                }
+            })
+        }
```

Comment:
I think the approach at https://github.com/astral-sh/uv/pull/1786/commits/51b38f9a871c42ac6638328718c04630b216e56f#diff-e0d9d19e8dbea69ecd5354dcbec7d87e259e801687037360fac463c3143c16c8R1004-R1011 is probably a better way to handle the "branch or tag" refspecs.

Reply:
Picked this out of https://github.com/astral-sh/uv/pull/1786 in https://github.com/astral-sh/uv/pull/1781/commits/afafb66b34c604c3ba274ee402083973a953d77a